### DISCUSSION TITLE
cross-only build with no native compiler installed?

### DISCUSSION THREAD

**Discussion by LunarLambda** [mesonbuild/meson#11249] [Q&A] [answered]
I'm working on a meson-based toolchain for developing Game Boy Advance games.

The homebrew community has many less tech-savvy users, and "installing a native compiler" can actually be a fairly difficult step for many of them, particularly on Windows or Mac where Visual Studio and Xcode can be huge downloads (7 GB for MSVC and the Windows 10 SDK!)

Does meson support a "cross-only" build if you specify a cross file and nothing in your code uses `native: true`? Or would you have to lie to meson by using the cross file as a native file, or similar?

If it's not supported, would it be possible to have it be supported? It's a somewhat fringe use case, but there are many projects that only need a target compiler, and it could also speed up things like CI.

**Comment by eli-schwartz** [comment, marked as answer]
Listing a language in the `project()` function is similar to invoking the `add_languages()` function with `required: true`.

In both cases, Meson will attempt to detect a build machine compiler (native) as well as a host machine compiler (cross) -- if the build is a native build, they're the same thing.

If the host machine compiler isn't found (the cross one in a cross build, the only one in a native build) then the value of "required" determines whether to error out (defaults to true). If the build machine (native) in a cross build is not found, it's ignored regardless of the required kwarg.

tl;dr -- this should work out of the box.

**Reply by eli-schwartz** [reply]
Note: you can specify `native: true` or `native: false` to the add_languages function, and it will then restrict itself to only checking for one, not both. In that case, the required kwarg applies to whichever one you chose, even if that's the native one in a cross build.

This means you can defer adding the base language by invoking `add_languages('c', native: false)` instead of listing it in project(), and meson won't even bother to check for the native compiler. This is a bit faster since less work done, and logs a bit less information, but shouldn't otherwise matter.

**Reply by LunarLambda** [reply]
awesome, thank you!
